otherlibs/unix/signals.c

```diff
@@ -69,22 +69,18 @@ CAMLprim value unix_sigprocmask(value vaction, value vset)
   how = sigprocmask_cmd[Int_val(vaction)];
   decode_sigset(vset, &set);
   caml_enter_blocking_section();
-  retcode = caml_sigmask_hook(how, &set, &oldset);
+  retcode = sigprocmask(how, &set, &oldset);
   caml_leave_blocking_section();
   /* Run any handlers for just-unmasked pending signals */
-  caml_process_pending_actions();
+  caml_process_pending_signals();
   if (retcode != 0) unix_error(retcode, "sigprocmask", Nothing);
   return encode_sigset(&oldset);
 }
 
 CAMLprim value unix_sigpending(value unit)
 {
   sigset_t pending;
-  int i;
   if (sigpending(&pending) == -1) uerror("sigpending", Nothing);
-  for (i = 1; i < NSIG; i++)
```

Comment:
Something has gone wrong here. We are not returning pending signals.

Reply:
Well, the code in this PR looks like the code in trunk before 2019: it's returning whatever `sigpending` says is pending, but not the signals that have already been received and recorded in `caml_pending_signals`.

This was fixed on trunk in 2019, commit 08ed1e85cd .  It would be nice to follow trunk here.  Just remember to use atomic reads to access `caml_pending_signals`.

Reply:
Fixed in https://github.com/ocaml-multicore/ocaml/commit/7593d3d0ad3780f494a50ec276308cde17613bf7